### ISSUE TITLE
#ask queries don't work when $wgReadOnly is set

### ISSUE THREAD

**Issue by hexmode** [SemanticMediaWiki/SemanticMediaWiki#3327]
### Setup and configuration

- SMW version: 2.5.6 (33df15e)
- MW version: 1.27.3 (ba7e9a1)
- PHP version: 5.6.32 (apache2handler)
- DB system (MySQL, Blazegraph, etc.) and version: 5.6.10-enterprise-commercial-advanced-log

### Issue
Any(?) #ask queries on a page while the wiki is in ReadOnly mode will end in a ReadOnlyError.

Changing 
```
$writeConnection = $this->writeConnection();
```
to
```
$writeConnection = $this->readConnection();
```
in [Database.php](https://github.com/SemanticMediaWiki/SemanticMediaWiki/blob/2.5.6/src/MediaWiki/Database.php#L289) fixes this.

**Comment by mwjames** [comment]
It should be obvious that setting the `$writeConnection` to a `$this->readConnection();` isn't the correct thing do to.

I never used `$wgReadOnly` so I do not really know why the supposedly `readConnection` should work since all select queries use the `read` and not the `write` connection. For `query` queries, a `writeConnection` is necessary due to auto-commit support and the need for temporary tables to create and park results while resolving hierarchies etc. . 

 Why does a wiki set `$wgReadOnly`? Which queries doesn't work?

**Comment by mwjames** [comment]
@kghbln Did `wgReadOnly` ever came up in any discussion? or on wiki?

@hexmode If `wgReadOnly` is temporary and you want write to use a read maybe changing `smwgLocalConnectionConf` will accomplish the same without changing the semantics of `$writeConnection = $this->readConnection();` in the `Database` class.

I did a quick test on setting `$wgReadOnly = true;` and queries are executed (using SMW 3.0) without any `ReadOnlyError` on MW 1.30.

**Comment by kghbln** [comment]
> @kghbln Did wgReadOnly ever came up in any discussion? or on wiki?

No, and I cannot replicate the issue either. The existing queries are accessible and still work and now one can obviously not be added.

**Setup**
- MediaWiki | 1.27.4 (536f392)23:56, 10. Jul. 2018
- PHP | 5.6.36-0+deb8u1 (apache2handler)
- MySQL | 5.5.60-0+deb8u1
- Semantic MediaWiki | 2.5.6 (33df15e) 14:09, 14. Feb. 2018

Ad hoc I see two cases where one would like to set `$wgReadOnly`
1. For temporary maintenance on private or public wikis - probably not big of an issue
2. For temporary absence (several days or weeks included) of the landlord on public wikis - this would be an issue

Still, things worked for my test.

**Comment by hexmode** [comment]
@mwjames writes:

> It should be obvious that setting the `$writeConnection` to a `$this->readConnection();` isn't the correct thing do to.

Yes, it seems pretty clear that this is not the right thing to do. Which is why I submitted a bug report instead of a pull request. :wink:
@mwjames writes:

> It should be obvious that setting the `$writeConnection` to a
> `$this->readConnection();` isn't the correct thing do to.

Yes, it seems pretty clear that this is not the right thing to do.
Which is why I submitted a bug report instead of a pull request. ;)

> For `query` queries, a `writeConnection` is necessary due to
> auto-commit support and the need for temporary tables to create and
> park results while resolving hierarchies etc. .

I'm not sure how temporary tables interact with the database during
$wgReadOnly.  I will try to get some input from people who know more
about MW internals.

From your explanation of why `writeConnection` is used and how I
understand `$wgReadOnly` (which could be completely wrong), I don't
think the `$wgReadOnly` should cause this issue since (in my
understanding) temporary tables wouldn't affect the state of the
database or the wiki.

>  Why does a wiki set `$wgReadOnly`?

In this particular case, I want to expose a read-only version of the
wiki to a web-crawler.  The wiki otherwise requires SSO authentication
via SAML to view and edit it.

Since the non-authenticated face of the wiki is only meant to be used by
the web crawler, I put it in read-only mode so that, even if someone
else gets access to the crawler's interface to the wiki, they won't be
able to modify anything on it.

>  Which queries doesn't work?

The following simple query on [[Sandbox]] shows a lot of results on the
RW face of the wiki but causes the ReadOnlyError on the RO face:

```
{{#ask:
[[Category:Content]]
}}
```
This category is the "top-level" that only contains other categories, no
pages.  It has 51 categories as members and some of those categories
have subcategories.

Note that the problem seems to be related to the size of the category.
If I change it to [[Category:Abbreviations]], the RO site does not show
an error.  The Abbreviations category contains only 24 pages and its
"parent" category is [[Category:Content]]

[[Sandbox test 2]] shows "4711" (or ReadOnlyError on the RO site) and is
set to
```
{{#expr:{{#arraymaptemplate:{{#ask:
[[Category:Content]]
|link=none
|sep=;
|limit=10000
}}|Sum|;|+}}}}
```
while [[Template:Sum]] is
```
{{#expr:{{#ask:
[[Category:{{PAGENAME:{{{1|}}}}}]]
|format=count
}}}}
```

Using the Special:Ask on the RO site so that the query string is trimmed
to just `?q=[[Category%3AContent]]` produces the ReadOnlyError.

> For `query` queries, a `writeConnection` is necessary due to auto-commit support and the need for temporary tables to create and park results while resolving hierarchies etc. .

I'm not sure how temporary tables interact with the database during `$wgReadOnly`.  I will try to get some input from people who know more about MW internals.

From your explanation of why `writeConnection` is used and how I understand `$wgReadOnly` (which could be completely wrong), I don't think the `$wgReadOnly` should cause this issue since (in my understanding) temporary tables wouldn't affect the state of the database or the wiki.

>  Why does a wiki set `$wgReadOnly`?

In this particular case, I want to expose a read-only version of the wiki to a web-crawler.  The wiki otherwise requires SSO authentication via SAML to view and edit it.

Since the non-authenticated face of the wiki is only meant to be used by the web crawler, I put it in read-only mode so that, even if someone else gets access to the crawler's interface to the wiki, they won't be able to modify anything on it.

>  Which queries doesn't work?

The following simple query on [[Sandbox]] shows a lot of results on the RW face of the wiki but causes the ReadOnlyError on the RO face:

```
{{#ask:
[[Category:Content]]
}}
```
This category is the "top-level" that only contains other categories, no pages.  It has 51 categories as members and some of those categories have subcategories.

Note that the problem seems to be related to the size of the category. If I change it to [[Category:Abbreviations]], the RO site does not show an error.  The Abbreviations category contains only 24 pages and its "parent" category is [[Category:Content]]

[[Sandbox test 2]] shows "4711" (or ReadOnlyError on the RO site) and is set to
```
{{#expr:{{#arraymaptemplate:{{#ask:
[[Category:Content]]
|link=none
|sep=;
|limit=10000
}}|Sum|;|+}}}}
```
while [[Template:Sum]] is
```
{{#expr:{{#ask:
[[Category:{{PAGENAME:{{{1|}}}}}]]
|format=count
}}}}
```
Finally, using the Special:Ask on the RO site so that the query string is trimmed to just `?q=[[Category%3AContent]]` produces the ReadOnlyError.

**Comment by mwjames** [comment]
> The following simple query on [[Sandbox]] shows a lot of results on the
RW face of the wiki but causes the ReadOnlyError on the RO face:

So, basically what this issue is about is a service (meaning non human access via API or other means) to access the wiki content?

As I said (and as @kghbln confirmed), setting `$wgReadOnly = true;` will allow you to use the wiki and query results from SMW without any issue and I currently cannot see how you would get a `ReadOnlyError` when browse/query results from within wiki.

* That said, I'm a bit fuzzy about your use case, the involvement of `wgReadOnly` and reports of `ReadOnlyError` (which I'm unable to produce by normal means).
* Looking at [0] there aren't many places that throw a `ReadOnlyError`.
* Try to set the `smwgLocalConnectionConf` to only use `DB_SLAVE` in case that solves your issue with `wgReadOnly` read connections (I still don't understand the relation here!).

PS: What does RW/RO face mean?

[0] https://github.com/wikimedia/mediawiki/search?l=PHP&p=1&q=ReadOnlyError

**Comment by hexmode** [comment]
> PS: What does RW/RO face stands for?
* **RO Face**: the **R**ead-**O**nly web inter**Face** to the wiki
* **RW Face**: the **R**ead/**W**rite web inter**Face** to the wiki

(Sorry, they made sense in my head. :stuck_out_tongue_closed_eyes: )

**Comment by mwjames** [comment]
> Could this be caused by the size of the data sets? Maybe I can find a way to get you a dataset that reproduces this.

No, because if you observe a `ReadOnlyError` then this is being thrown by a process unrelated to any data or its size.

> The problem has cleared up.

While the issue has been resolved by setting `smwgLocalConnectionConf`, I'm somewhat unsatisfied by the analysis of what causes the `ReadOnlyError` and why `DB_SLAVE` will remedy those errors.

**Comment by hexmode** [comment]
mwjames <notifications@github.com> writes:

> ... if you observe a `ReadOnlyError` then this is being thrown  by a process unrelated to any data or its size.
...
> While the issue has been resolved by setting `smwgLocalConnectionConf`, I'm somewhat unsatisfied by the analysis of what causes the `ReadOnlyError` and why `DB_SLAVE` will remedy those errors.

So, more work is needed to track this down.

reopening

**Comment by hexmode** [comment]
Backtrace
```
#0 .../extensions/SemanticMediaWiki/src/MediaWiki/Database.php(308): DatabaseBase->query('CREATE TEMPORAR...', 'SMW\\SQLStore\\Ta...', false)
#1 .../extensions/SemanticMediaWiki/src/SQLStore/TableBuilder/TemporaryTableBuilder.php(51): SMW\MediaWiki\Database->query('CREATE TEMPORAR...', 'SMW\\SQLStore\\Ta...', false, false)
#2 .../extensions/SemanticMediaWiki/src/SQLStore/QueryEngine/HierarchyTempTableBuilder.php(116): SMW\SQLStore\TableBuilder\TemporaryTableBuilder->create('`wiki1t2`')
#3 .../extensions/SemanticMediaWiki/src/SQLStore/QueryEngine/QuerySegmentListProcessor.php(301): SMW\SQLStore\QueryEngine\HierarchyTempTableBuilder->createHierarchyTempTableFor('class', '`wiki1t2`', '('31904')')
#4 .../extensions/SemanticMediaWiki/src/SQLStore/QueryEngine/QuerySegmentListProcessor.php(239): SMW\SQLStore\QueryEngine\QuerySegmentListProcessor->resolveHierarchyForSegment(Object(SMW\SQLStore\QueryEngine\QuerySegment))
#5 .../extensions/SemanticMediaWiki/src/SQLStore/QueryEngine/QuerySegmentListProcessor.php(123): SMW\SQLStore\QueryEngine\QuerySegmentListProcessor->doResolveBySegment(Object(SMW\SQLStore\QueryEngine\QuerySegment))
#6 .../extensions/SemanticMediaWiki/src/SQLStore/QueryEngine/QuerySegmentListProcessor.php(123): SMW\SQLStore\QueryEngine\QuerySegmentListProcessor->doResolveBySegment(Object(SMW\SQLStore\QueryEngine\QuerySegment))
#7 .../extensions/SemanticMediaWiki/src/SQLStore/QueryEngine/QuerySegmentListProcessor.php(112): SMW\SQLStore\QueryEngine\QuerySegmentListProcessor->doResolveBySegment(Object(SMW\SQLStore\QueryEngine\QuerySegment))
#8 .../extensions/SemanticMediaWiki/src/SQLStore/QueryEngine/QueryEngine.php(203): SMW\SQLStore\QueryEngine\QuerySegmentListProcessor->doResolveQueryDependenciesById(0)
#9 .../extensions/SemanticMediaWiki/includes/storage/SQLStore/SMW_SQLStore3.php(330): SMW\SQLStore\QueryEngine\QueryEngine->getQueryResult(Object(SMWQuery))
#10 .../extensions/SemanticMediaWiki/includes/storage/SQLStore/SMW_SQLStore3.php(318): SMWSQLStore3->fetchQueryResult(Object(SMWQuery))
#11 .../extensions/SemanticMediaWiki/includes/query/SMW_QueryProcessor.php(416): SMWSQLStore3->getQueryResult(Object(SMWQuery))
#12 .../extensions/SemanticMediaWiki/src/ParserFunctions/AskParserFunction.php(211): SMWQueryProcessor::getResultFromQuery(Object(SMWQuery), Array, 2, 1001)
#13 .../extensions/SemanticMediaWiki/src/ParserFunctions/AskParserFunction.php(136): SMW\ParserFunctions\AskParserFunction->doFetchResultsFromFunctionParameters(Array)
#14 .../extensions/SemanticMediaWiki/src/ParserFunctionFactory.php(298): SMW\ParserFunctions\AskParserFunction->parse(Array)
#15 [internal function]: SMW\ParserFunctionFactory->SMW\{closure}(Object(Parser), '[[Category:Cont...', 'link=none', 'sep=;', 'limit=1000')
#16 .../includes/parser/Parser.php(3834): call_user_func_array(Object(Closure), Array)
#17 .../includes/parser/Parser.php(3566): Parser->callParserFunction(Object(PPFrame_DOM), '#ask', Array)
#18 .../includes/parser/Preprocessor_DOM.php(1175): Parser->braceSubstitution(Array, Object(PPFrame_DOM))
#19 .../includes/parser/Parser.php(3484): PPFrame_DOM->expand(Object(PPNode_DOM))
#20 .../includes/parser/Preprocessor_DOM.php(1175): Parser->braceSubstitution(Array, Object(PPFrame_DOM))
#21 .../includes/parser/Parser.php(3484): PPFrame_DOM->expand(Object(PPNode_DOM))
#22 .../includes/parser/Preprocessor_DOM.php(1175): Parser->braceSubstitution(Array, Object(PPFrame_DOM))
#23 .../includes/parser/Parser.php(3380): PPFrame_DOM->expand(Object(PPNode_DOM), 0)
#24 .../includes/parser/Parser.php(1261): Parser->replaceVariables('{{#expr:{{#arra...')
#25 .../includes/parser/Parser.php(452): Parser->internalParse('{{#expr:{{#arra...')
#26 .../includes/content/WikitextContent.php(331): Parser->parse('{{#expr:{{#arra...', Object(Title), Object(ParserOptions), true, true, 201901)
#27 .../includes/content/AbstractContent.php(497): WikitextContent->fillParserOutput(Object(Title), 201901, Object(ParserOptions), true, Object(ParserOutput))
#28 .../includes/poolcounter/PoolWorkArticleView.php(140): AbstractContent->getParserOutput(Object(Title), 201901, Object(ParserOptions))
#29 .../includes/poolcounter/PoolCounterWork.php(123): PoolWorkArticleView->doWork()
#30 .../includes/page/Article.php(666): PoolCounterWork->execute()
#31 .../includes/actions/ViewAction.php(44): Article->view()
#32 .../includes/MediaWiki.php(493): ViewAction->show()
#33 .../includes/MediaWiki.php(288): MediaWiki->performAction(Object(Article), Object(Title))
#34 .../includes/MediaWiki.php(735): MediaWiki->performRequest()
#35 .../includes/MediaWiki.php(509): MediaWiki->main()
#36 .../index.php(43): MediaWiki->run()
#37 {main}
```

**Comment by hexmode** [comment]
Looks like [`newTemporaryTableBuilder`](https://github.com/SemanticMediaWiki/SemanticMediaWiki/blob/53e75cd043490cda999d6d8cdee352192f39dd92/src/SQLStore/QueryEngineFactory.php#L166) uses the connection provided by `mw.db.queryengine` now as the result of a [change you made two years ago](https://github.com/SemanticMediaWiki/SemanticMediaWiki/commit/3098edd9bdd27f896a68abb4a0c52e2338765dcb) to resolve [an unrelated(?) issue](https://github.com/SemanticMediaWiki/SemanticMediaWiki/issues/1757).

**Comment by mwjames** [comment]
>Looks like newTemporaryTableBuilder uses the connection provided by mw.db.queryengine now as the result of a change you made two years ago to resolve an unrelated(?) issue.

> Backtrace

The backtrace is what I wanted to see to identify the entry point of the issue.

> #0 .../extensions/SemanticMediaWiki/src/MediaWiki/Database.php(308): DatabaseBase->query('CREATE TEMPORAR...', 'SMW\\SQLStore\\Ta...', false)
#1 .../extensions/SemanticMediaWiki/src/SQLStore/TableBuilder/TemporaryTableBuilder.php(51): SMW\MediaWiki\Database->query('CREATE TEMPORAR...', 'SMW\\SQLStore\\Ta...', false, false)

As I thought, creating temporary tables is most likely in the center of the cause of this issue (the cited PR is unrelated, as it mitigates the replication in case of MySQL GTID) and the call to `DatabaseBase->query('CREATE TEMPORAR...', 'SMW\\SQLStore\\Ta...', false)` proves what should actually be allowed while being in `wgReadOnly` on DB_MASTER but apparently it is not, so setting the `smwgLocalConnectionConf` to DB_SLAVE avoids your `ReadOnlyError`.

You may want to investigate (in MW core):
* Why temp table creation in  `wgReadOnly`  (see comment https://github.com/SemanticMediaWiki/SemanticMediaWiki/pull/2534#issuecomment-325649500) causes `ReadOnlyError`
* Whether this can be fixed in MW core or not (it may that this has already been fixed with MW 1.30+)

As for SMW, the documentation states that we expect temporary table creation to work (using the `ElasticStore` or `SPARQLStore` doesn't require it) and while the intent for `smwgLocalConnectionConf` was different, if it can be used with `DB_SLAVE` on both `read` and `write` to circumvent the reported issue then in fact this issue is solved.

@kghbln FYI

**Comment by kghbln** [comment]
Great. Thanks a lot for investigating into the abyss of the issue thus tracking the cause.

> @kghbln FYI

I referenced this issue at `$smwgLocalConnectionConf`

> You may want to investigate (in MW core):
>
> -   Why temp table creation in wgReadOnly (see comment #2534 (comment)) causes ReadOnlyError
> -  Whether this can be fixed in MW core or not (it may that this has already been fixed with MW 1.30+)

It will be great if the follow-up Phabricator task will be referenced here if one gets created. I am closing this since I do not think there is currently something to do for SMW.

**Comment by hexmode** [comment]
This [note in the Database::query()](https://github.com/wikimedia/mediawiki/blob/130ec2523df12a3ca2fe0d422163696d09fcea08/includes/libs/rdbms/database/Database.php#L1109) probably explains it:
``` php
# In theory, non-persistent writes are allowed in read-only mode, but due to things
# like https://bugs.mysql.com/bug.php?id=33669 that might not work anyway...
$reason = $this->getReadOnlyReason();
if ( $reason !== false ) {
	throw new DBReadOnlyError( $this, "Database is read-only: $reason" );
}
```

[MySQL Bug #33669](https://bugs.mysql.com/bug.php?id=33669) ("Transactional temporary tables do not work under --read-only") is closed as fixed.

[Filed a bug](https://phabricator.wikimedia.org/T202245).
